File: TSCompiler/Semantic/RTL/undefined.cpp
#include "../semantic.h"

ClassDeclarationNode *Semantic::createUndefinedClass()
{
    auto *body = ClassElementListNode::makeEmpty();

    {
        auto *clsConstructor = new ClassElementNode(
            RequiredParameterListNode::makeEmpty(), nullptr);
        body->add(clsConstructor);
        clsConstructor->propertyAndReturnType =
            new TypeNode(new JvmDataType(JvmDataType::Type::Void));
    }

    // Methods

    // Plus

    std::vector<JvmDataType> types = {RTL_NULL_TYPE,    RTL_NUMBER_TYPE,
                                      RTL_BOOLEAN_TYPE, RTL_STRING_TYPE,
                                      RTL_VOID_TYPE,    RTL_UNDEFINED_TYPE};

    {
        auto *type = new TypeNode(new JvmDataType(RTL_STRING_TYPE));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "plus", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_STRING_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    for (const auto &jvmType : types)
    {
        if (jvmType == RTL_STRING_TYPE ||
            (jvmType == RTL_UNDEFINED_TYPE && jvmType != RTL_VOID_TYPE))
            continue;

        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "plus", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_NUMBER_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // Minus

    std::vector<JvmDataType> operationTypes = {RTL_NUMBER_TYPE, RTL_NULL_TYPE,
                                               RTL_UNDEFINED_TYPE};

    for (const auto &jvmType : operationTypes)
    {
        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "minus", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_NUMBER_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // Mul

    for (const auto &jvmType : operationTypes)
    {
        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "mul", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_NUMBER_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // Div

    for (const auto &jvmType : operationTypes)
    {
        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "div", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_NUMBER_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    {
        auto *method = new ClassElementNode(
            "uPlus", RequiredParameterListNode::makeEmpty(),
            new TypeNode(new JvmDataType(RTL_NUMBER_TYPE)), nullptr);
        body->add(method);
    }
    {
        auto *method = new ClassElementNode(
            "uMinus", RequiredParameterListNode::makeEmpty(),
            new TypeNode(new JvmDataType(RTL_NUMBER_TYPE)), nullptr);
        body->add(method);
    }

    // Less

    for (const auto &jvmType : types)
    {
        if (jvmType == RTL_UNDEFINED_TYPE && jvmType != RTL_VOID_TYPE)
            continue;

        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "less", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // Great

    for (const auto &jvmType : types)
    {
        if (jvmType == RTL_UNDEFINED_TYPE && jvmType != RTL_VOID_TYPE)
            continue;

        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "great", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // Equal

    for (const auto &jvmType : types)
    {
        if (jvmType == RTL_UNDEFINED_TYPE && jvmType != RTL_VOID_TYPE)
            continue;

        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "equal", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // NotEqual

    for (const auto &jvmType : types)
    {
        if (jvmType == RTL_UNDEFINED_TYPE && jvmType != RTL_VOID_TYPE)
            continue;

        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "notEqual", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    // LessEqual

    for (const auto &jvmType : types)
    {
        if (jvmType == RTL_UNDEFINED_TYPE && jvmType != RTL_VOID_TYPE)
            continue;

        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "lessEqual", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    for (const auto &jvmType : types)
    {
        if (jvmType == RTL_UNDEFINED_TYPE && jvmType != RTL_VOID_TYPE)
            continue;

        auto *type = new TypeNode(new JvmDataType(jvmType));
        auto *param = new RequiredParameterNode("other", type);
        auto *method = new ClassElementNode(
            "greatEqual", new RequiredParameterListNode(param),
            new TypeNode(new JvmDataType(RTL_BOOLEAN_TYPE)), nullptr);
        method->analyzeArguments();
        body->add(method);
    }

    auto *rtlClass = new ClassDeclarationNode("JavaRTL/Undefined", body);
    for (auto *elem : body->GetSeq()) elem->elemClass = rtlClass;
    return rtlClass;
}

File: TSCompiler/JavaRTL/Undefined.java
package JavaRTL;

public class Undefined extends JavaRTL.Void {
    public JavaRTL.Number plus(JavaRTL.Null other) {
        return new JavaRTL.Number(this);
    }
    public JavaRTL.Number plus(JavaRTL.Number other) {
        return new JavaRTL.Number(this);
    }
    public JavaRTL.String plus(JavaRTL.String other) {
        return new JavaRTL.String(this).plus(other);
    }
    public JavaRTL.Number plus(JavaRTL.Boolean other) {
        return new JavaRTL.Number(this);
    }
    public JavaRTL.Number plus(JavaRTL.Void other) {
        return new JavaRTL.Number(this);
    }

    public JavaRTL.Number minus(JavaRTL.Null other) {
        return new JavaRTL.Number(this);
    }
    public JavaRTL.Number minus(JavaRTL.Number other) {
        return new JavaRTL.Number(this);
    }
    public JavaRTL.Number minus(JavaRTL.Undefined other) {
        return new JavaRTL.Number(this);
    }

    public JavaRTL.Number mul(JavaRTL.Null other) {
        return new JavaRTL.Number(this);
    }
    public JavaRTL.Number mul(JavaRTL.Number other) {
        return new JavaRTL.Number(this);
    }
    public JavaRTL.Number mul(JavaRTL.Undefined other) {
        return new JavaRTL.Number(this);
    }

    public JavaRTL.Number div(JavaRTL.Null other) {
        return new JavaRTL.Number(this);
    }
    public JavaRTL.Number div(JavaRTL.Number other) {
        return new JavaRTL.Number(this);
    }
    public JavaRTL.Number div(JavaRTL.Undefined other) {
        return new JavaRTL.Number(this);
    }

    public JavaRTL.Boolean less(JavaRTL.Null other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean less(JavaRTL.Number other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean less(JavaRTL.String other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean less(JavaRTL.Boolean other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean less(JavaRTL.Void other) {
        return new JavaRTL.Boolean(false);
    }

    public JavaRTL.Boolean great(JavaRTL.Null other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean great(JavaRTL.Number other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean great(JavaRTL.String other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean great(JavaRTL.Boolean other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean great(JavaRTL.Void other) {
        return new JavaRTL.Boolean(false);
    }

    public JavaRTL.Boolean equal(JavaRTL.Null other) {
        return new JavaRTL.Boolean(true);
    }
    public JavaRTL.Boolean equal(JavaRTL.Number other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean equal(JavaRTL.String other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean equal(JavaRTL.Boolean other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean equal(JavaRTL.Void other) {
        return new JavaRTL.Boolean(true);
    }

    // that method not adding into ASI
    public JavaRTL.Boolean equal(Any other) {
        if (other instanceof JavaRTL.Void || other instanceof JavaRTL.Null) {
            return new JavaRTL.Boolean(true);
        }
        return super.equal(other);
    }

    public JavaRTL.Boolean notEqual(JavaRTL.Null other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean notEqual(JavaRTL.Number other) {
        return new JavaRTL.Boolean(true);
    }
    public JavaRTL.Boolean notEqual(JavaRTL.String other) {
        return new JavaRTL.Boolean(true);
    }
    public JavaRTL.Boolean notEqual(JavaRTL.Boolean other) {
        return new JavaRTL.Boolean(true);
    }
    public JavaRTL.Boolean notEqual(JavaRTL.Void other) {
        return new JavaRTL.Boolean(false);
    }

    // that method not adding into ASI
    public JavaRTL.Boolean notEqual(Any other) {
        if (other instanceof JavaRTL.Void || other instanceof JavaRTL.Null) {
            return new JavaRTL.Boolean(false);
        }
        return super.notEqual(other);
    }

    public JavaRTL.Boolean lessEqual(JavaRTL.Null other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean lessEqual(JavaRTL.Number other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean lessEqual(JavaRTL.String other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean lessEqual(JavaRTL.Boolean other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean lessEqual(JavaRTL.Void other) {
        return new JavaRTL.Boolean(false);
    }

    public JavaRTL.Boolean greatEqual(JavaRTL.Null other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean greatEqual(JavaRTL.Number other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean greatEqual(JavaRTL.String other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean greatEqual(JavaRTL.Boolean other) {
        return new JavaRTL.Boolean(false);
    }
    public JavaRTL.Boolean greatEqual(JavaRTL.Void other) {
        return new JavaRTL.Boolean(false);
    }

    @Override
    public java.lang.String toString() {
        return "undefined";
    }
}
